pkg/volume/gce_pd/gce_util.go

```diff
@@ -122,6 +125,167 @@ func (util *GCEDiskUtil) DetachDisk(pd *gcePersistentDisk) error {
 	if err := gce.(*gce_cloud.GCECloud).DetachDisk(pd.pdName); err != nil {
 		return err
 	}
+
+	// Verify disk detached, retry if needed.
+	go verifyDetached(pd, gce)
+	return nil
+}
+
+// Verifys the disk device to be created has been succesffully attached, and retries if it fails.
+func verifyAttached(pd *gcePersistentDisk, sdBeforeSet util.StringSet, gce cloudprovider.Interface) (string, error) {
+	devicePaths := getDiskByIdPaths(pd)
+	for numRetries := 0; numRetries < maxRetries; numRetries++ {
+		for numChecks := 0; numChecks < maxChecks; numChecks++ {
+			if err := udevadmChangeToNewDrives(sdBeforeSet); err != nil {
+				// udevadm errors should not block disk attachment, log and continue
+				glog.Errorf("%v", err)
+			}
+
+			for _, path := range devicePaths {
+				if pathExists, err := pathExists(path); err != nil {
+					return "", err
+				} else if pathExists {
+					// A device path has succesfully been created for the PD
+					glog.V(5).Infof("Succesfully attached GCE PD %q.", pd.pdName)
+					return path, nil
+				}
+			}
+
+			// Sleep then check again
+			glog.V(5).Infof("Waiting for GCE PD %q to attach.", pd.pdName)
+			time.Sleep(checkSleepDuration)
+		}
+
+		// Try attaching the disk again
+		glog.Warningf("Timed out waiting for GCE PD %q to attach. Retrying attach.", pd.pdName)
+		if err := gce.(*gce_cloud.GCECloud).AttachDisk(pd.pdName, pd.readOnly); err != nil {
+			return "", err
+		}
+	}
+
+	return "", fmt.Errorf("Could not attach GCE PD %q. Timeout waiting for mount paths to be created.", pd.pdName)
+}
+
+// Veify the specified persistent disk device has been succesfully detached, and retries if it fails.
+func verifyDetached(pd *gcePersistentDisk, gce cloudprovider.Interface) {
+	ch := detachChanManager.CreateAndAddChan(pd.pdName, 0 /* bufferSize */)
```

Comment:
Please add a comment to the effect:

bufferSize being 0 is very important, because it means that when senders send, they are blocked until we recieve; this avoids having to have a separate "did you exit yet" check.